### PR TITLE
Use dav versions in files

### PR DESCRIPTION
**Pull request by DeepDiver1975**
## Description
This PR is based on #29207 and integrated files_versions ui via webdav

## Related Issue
#29087 

## Motivation and Context
We shall use the webdav api for versioning

## How Has This Been Tested?
- manual testing in files app
- cadaver

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: apps/dav/lib/DAV/CorePlugin.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+use OCA\DAV\Connector\Sabre\File;
+use OCA\DAV\Files\ICopySource;
+use Sabre\DAV\IFile;
+use Sabre\DAV\Server;
+use Sabre\DAV\ServerPlugin;
+use Sabre\HTTP\RequestInterface;
+use Sabre\HTTP\ResponseInterface;
+
+class CorePlugin extends ServerPlugin {
```
Comment (PVince81) [inline apps/dav/lib/DAV/CorePlugin.php]: please rename to something that says "plugin to avoid deletion on copy" or something. `CorePlugin` is the wrong name for this. Also please add PHPDoc to describe this.

File: apps/dav/lib/DAV/CorePlugin.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+			return true;
+		}
+
+		if (!$this->server->emit('beforeBind', [$copyInfo['destination']])) return false;
+
+		$copySuccess = false;
+		if ($sourceNode instanceof ICopySource) {
+			$copySuccess = $sourceNode->copy($destinationNode->getFileInfo()->getPath());
+		}
+		if (!$copySuccess) {
```
Comment (PVince81) [inline apps/dav/lib/DAV/CorePlugin.php]: wouldn't doing a simple `return` here let the actual CorePlugin do the fallback ? this would remove the need for duplicating the response related code below

File: apps/dav/lib/Files/ICopySource.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+
+namespace OCA\DAV\Files;
+
+
+interface ICopySource {
```
Comment (PVince81) [inline apps/dav/lib/Files/ICopySource.php]: PHPDoc

File: apps/dav/lib/Files/ICopySource.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+	/**
+	 * @param string $destinationPath
+	 * @return boolean
+	 */
+	public function copy($destinationPath);
```
Comment (PVince81) [inline apps/dav/lib/Files/ICopySource.php]: Write a short sentence for the PHPDoc, at least on the interface.

File: apps/dav/lib/Meta/MetaFile.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+
+namespace OCA\DAV\Meta;
+
+
+use OC\Files\Meta\MetaFileVersionNode;
+use OCA\DAV\Files\ICopySource;
+use Sabre\DAV\File;
+
+class MetaFile extends File implements ICopySource {
```
Comment (PVince81) [inline apps/dav/lib/Meta/MetaFile.php]: what's a MetaFile ? => PHPDoc

File: apps/dav/lib/Meta/MetaFolder.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+namespace OCA\DAV\Meta;
+
+
+use OCP\Files\File;
+use OCP\Files\Folder;
+use OCP\Files\Node;
+use Sabre\DAV\Collection;
+
+class MetaFolder extends Collection {
```
Comment (PVince81) [inline apps/dav/lib/Meta/MetaFolder.php]: same. Please check PHPDoc of all new classes.

File: apps/dav/lib/Meta/MetaFolder.php [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	function getName() {
+		return $this->folder->getName();
+	}
+
+	public static function nodeFactory(Node $node) {
```
Comment (PVince81) [inline apps/dav/lib/Meta/MetaFolder.php]: can this be made private ?

File: apps/dav/lib/Meta/RootCollection.php [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+		} catch (NotFoundException $exception) {
+			throw new NotFound();
+		}
+	}
+	
+	/**
+	 * @inheritdoc
+	 */
+	function getChildren() {
+		throw new MethodNotAllowed('Listing members of this collection is disabled');
```
Comment (PVince81) [inline apps/dav/lib/Meta/RootCollection.php]: => `is not allowed`. Disabled sounds like there is a way to enable it with a switch.

File: apps/dav/lib/RootCollection.php [hunk 9]
```diff
@@ -106,7 +106,8 @@ public function __construct() {
 				$systemTagCollection,
 				$systemTagRelationsCollection,
 				$uploadCollection,
-				$avatarCollection
+				$avatarCollection,
+				new \OCA\DAV\Meta\RootCollection(\OC::$server->getRootFolder())
```
Comment (PVince81) [inline apps/dav/lib/RootCollection.php]: inject IRootFolder ? And in the location where you inject, I suggest using `\OC::$server->getLazyRootFolder()` if that makes sense

File: apps/federation/tests/BackgroundJob/GetSharedSecretTest.php [hunk 10]
```diff
@@ -35,6 +34,7 @@
 use OCP\Http\Client\IResponse;
 use OCP\ILogger;
 use OCP\IURLGenerator;
+use Test\TestCase;
```
Comment (PVince81) [inline apps/federation/tests/BackgroundJob/GetSharedSecretTest.php]: is this change related in any way ?

File: apps/files_versions/appinfo/routes.php [hunk 11]
```diff
@@ -35,10 +35,3 @@ function() {
 	require_once __DIR__ . '/../ajax/preview.php';
 });
 
-$this->create('files_versions_download', 'download.php')
-	->actionInclude('files_versions/download.php');
-$this->create('files_versions_ajax_getVersions', 'ajax/getVersions.php')
-	->actionInclude('files_versions/ajax/getVersions.php');
-$this->create('files_versions_ajax_rollbackVersion', 'ajax/rollbackVersion.php')
```
Comment (PVince81) [inline apps/files_versions/appinfo/routes.php]: tick the boxes in https://github.com/owncloud/core/issues/13007 after merging

File: apps/files_versions/lib/Storage.php [hunk 12]
```diff
@@ -461,6 +461,7 @@ public static function getVersions($uid, $filename, $userFullPath = '') {
 						$versions[$key]['path'] = Filesystem::normalizePath($pathinfo['dirname'] . '/' . $filename);
 						$versions[$key]['name'] = $versionedFile;
 						$versions[$key]['size'] = $view->filesize($dir . '/' . $entryName);
+						$versions[$key]['storage_location'] = "$dir/$entryName";
```
Comment (PVince81) [inline apps/files_versions/lib/Storage.php]: assuming this worked well with external storages also ?

File: apps/files_versions/tests/js/versioncollectionSpec.js [hunk 13]
```diff
@@ -1,161 +0,0 @@
-/*
- * Copyright (c) 2015
- *
- * This file is licensed under the Affero General Public License version 3
- * or later.
- *
- * See the COPYING-README file.
- *
- */
-describe('OCA.Versions.VersionCollection', function() {
```
Comment (PVince81) [inline apps/files_versions/tests/js/versioncollectionSpec.js]: no more tests for the collection ?

File: core/js/files/client.js [hunk 14]
```diff
@@ -737,40 +783,23 @@
...
+		 * Copies path to another path
+		 *
+		 * @param {String} path path to copy
+		 * @param {String} destinationPath destination path
+		 * @param {boolean} [allowOverwrite=false] true to allow overwriting,
+		 * false otherwise
+		 * @param {Object} [headers=null] additional headers
+		 *
+		 * @return {Promise} promise
+		 */
```
Comment (PVince81) [inline core/js/files/client.js]: `@since 10.0.4` ?

File: lib/private/Files/Meta/MetaFileVersionNode.php [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+		return $this->versionId;
+	}
+
+	public function getContent() {
+		return $this->storage->getContentOfVersion($this->internalPath, $this->versionId);
+	}
+
+	public function copy($targetPath) {
+		//TODO: inject
+		$target = \OC::$server->getRootFolder()->get($targetPath);
```
Comment (PVince81) [inline lib/private/Files/Meta/MetaFileVersionNode.php]: can we inject the root folder ? (lazy root folder if possible)

File: apps/dav/lib/DAV/CorePlugin.php [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+		$destinationNode = $copyInfo['destinationNode'];
+		if (!$copyInfo['destinationExists'] || !$destinationNode instanceof File || !$sourceNode instanceof IFile) {
+			return true;
+		}
+
+		if (!$this->server->emit('beforeBind', [$copyInfo['destination']])) return false;
+
+		$copySuccess = false;
+		if ($sourceNode instanceof ICopySource) {
+			$copySuccess = $sourceNode->copy($destinationNode->getFileInfo()->getPath());
```
Comment (PVince81) [inline apps/dav/lib/DAV/CorePlugin.php]: Do I understand it correctly that doing a Webdav COPY is not the same as calling `$storage->restoreVersion()` ?
If it was I'd expect to have some validation to make sure the COPY is operating on the same storage + target file and fail if it doesn't.

File: lib/private/Files/Meta/MetaVersionCollection.php [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+		$path = $view->getPath($this->fileId);
+		/** @var Storage $storage */
+		list($storage, $internalPath) = $view->resolvePath($path);
+		if (!$storage->instanceOfStorage(IVersionedStorage::class)) {
+			return [];
+		}
+		/** @var IVersionedStorage | Storage $storage */
+		$versions = $storage->getVersions($internalPath);
+		return array_map(function($version) use ($storage, $internalPath) {
+			return new MetaFileVersionNode($this, $version['version'],$storage, $internalPath);
```
Comment (PVince81) [inline lib/private/Files/Meta/MetaVersionCollection.php]: missing space `],$storage`

File: lib/public/Files/Storage/IVersionedStorage.php [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+ */
+
+
+namespace OCP\Files\Storage;
+
+/**
+ * Interface IVersionedStorage - storage layer to access version of a file
+ *
+ * @package OCP\Files\Storage
+ * @since 10.1.0
```
Comment (PVince81) [inline lib/public/Files/Storage/IVersionedStorage.php]: so no backports then?

**Review by PVince81 — COMMENTED** [review]
Looks good so far.

I didn't remember seeing tests for the new Sabre nodes, please add some. (I expect codecov will complain about this).

Can you clarify the part about Webdav COPY vs $storage->restoreVersion. If they are connected, does the error message propagate properly when restoring to another file than the original one ? (no HTTP 500).

File: apps/federation/tests/BackgroundJob/GetSharedSecretTest.php (hunk 10 above)
Comment (DeepDiver1975) [reply apps/federation/tests/BackgroundJob/GetSharedSecretTest.php]: fixing stuff while hitting it: Try to run federation tests with files_sharing dieabled -> 💥

File: apps/files_versions/lib/Storage.php (hunk 12 above)
Comment (DeepDiver1975) [reply apps/files_versions/lib/Storage.php]: Yes - tested with local mount

File: apps/files_versions/tests/js/versioncollectionSpec.js (hunk 13 above)
Comment (DeepDiver1975) [reply apps/files_versions/tests/js/versioncollectionSpec.js]: the collection is being tested with the model spec from my pov

File: lib/public/Files/Storage/IVersionedStorage.php (hunk 18 above)
Comment (DeepDiver1975) [reply lib/public/Files/Storage/IVersionedStorage.php]: we need to make a decision on this - since no database change is involved we can add this to 10.0.x as well

File: apps/files_versions/tests/js/versioncollectionSpec.js (hunk 13 above)
Comment (PVince81) [reply apps/files_versions/tests/js/versioncollectionSpec.js]: From what I see the collection still contains more complex logic related to pagination, which I believe was tested here. If the collection had simple or no logic I'd be fine, but it has much more than that.

By the way, does pagination still work with the DAV endpoint ? You might want to remove it (`fetchNext`, etc) if it isn't.

Comment (DeepDiver1975) [reply apps/files_versions/tests/js/versioncollectionSpec.js]: there is no pagination at the moment - I can give it a try and see if my js skills are enough to remove this stuff ....

File: apps/dav/lib/DAV/CorePlugin.php (hunk 2 above)
Comment (DeepDiver1975) [reply apps/dav/lib/DAV/CorePlugin.php]: No because the fall back will still perform the delete of the target.

File: apps/dav/lib/DAV/CorePlugin.php (hunk 16 above)
Comment (DeepDiver1975) [reply apps/dav/lib/DAV/CorePlugin.php]: > Do I understand it correctly that doing a Webdav COPY is not the same as calling $storage->restoreVersion() ?

within the implementation of copy() restoreVersion() is called if applicable

File: apps/dav/lib/Files/ICopySource.php (hunk 3 above)
Comment (DeepDiver1975) [reply apps/dav/lib/Files/ICopySource.php]: done

File: apps/dav/lib/Files/ICopySource.php (hunk 4 above)
Comment (DeepDiver1975) [reply apps/dav/lib/Files/ICopySource.php]: done

File: apps/dav/lib/Meta/MetaFile.php (hunk 5 above)
Comment (DeepDiver1975) [reply apps/dav/lib/Meta/MetaFile.php]: done

File: apps/dav/lib/Meta/MetaFolder.php (hunk 7 above)
Comment (DeepDiver1975) [reply apps/dav/lib/Meta/MetaFolder.php]: yes - done

File: apps/dav/lib/Meta/RootCollection.php (hunk 8 above)
Comment (DeepDiver1975) [reply apps/dav/lib/Meta/RootCollection.php]: will eventually be possible

File: apps/dav/lib/RootCollection.php (hunk 9 above)
Comment (DeepDiver1975) [reply apps/dav/lib/RootCollection.php]: there is no injection in this class at all

Comment (PVince81) [reply apps/dav/lib/RootCollection.php]: that's sad. when working on customgroups I went to great lengths (and pains) to inject and pass around injected members to RootCollection and every sub-sub-sub nodes. Anyway, fine by me for now.

**Comment by PVince81** [comment]
some known troubles with multiple  mounts: https://github.com/owncloud/core/issues/29490

**Comment by DeepDiver1975** [comment]
@PVince81 please review and test - THX

File: apps/files_versions/tests/js/versionstabviewSpec.js [hunk 19]
```diff
@@ -113,123 +113,4 @@ describe('OCA.Versions.VersionsTabView', function() {
...
-			revertStub = sinon.stub(VersionModel.prototype, 'revert');
-			tabView.setFileInfo(fileInfoModel);
-			tabView.collection.set(testVersions);
-		});
-		
-		afterEach(function() {
-			revertStub.restore();
-		});
-
-		it('tells the model to revert when clicking "Revert"', function() {
```
Comment (PVince81) [inline apps/files_versions/tests/js/versionstabviewSpec.js]: we're losing coverage here... please try and bring back many of those tests

**Comment by PVince81** [comment]
- [x] TEST: restore version
- [ ] BUG: download a version returns the file as a timestamp file name. Please adjust content-disposition header

**Comment by DeepDiver1975** [comment]
Move catch.  Thx

File: apps/files_versions/tests/js/versionstabviewSpec.js (hunk 19 above)
Comment (DeepDiver1975) [reply apps/files_versions/tests/js/versionstabviewSpec.js]: looks like I killed too much code ... only 'More button' should have been removed ....

**Comment by PVince81** [comment]
I'm now getting contradicting content-disposition headers:
```
  % curl -D - -X GET -u admin:admin -o temp.dat http://localhost/owncloud/remote.php/dav/meta/45/v/1510656635
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0HTTP/1.1 200 OK
Date: Tue, 14 Nov 2017 10:53:14 GMT
Server: Apache/2.4.29 (Linux/SUSE)
X-Powered-By: PHP/7.1.11
Set-Cookie: ockbfur1449k=l4ksce28t745hskgjp7onqb6g5; path=/owncloud; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate
Pragma: no-cache
Set-Cookie: oc_sessionPassphrase=mo0PEWR%2FHNZACVlimOgD4gCFseDHu5A7POP0TzF5upFnr9QlExZ73IE%2B6hJ%2F4F%2B%2BNXCj5r3McrAtfO2zxSCOTGmHqElQ32w0Xq4%2FRaG3iGB9s5UCT%2BGEfGNDuP31FGTT; path=/owncloud; HttpOnly
Content-Security-Policy: default-src 'none';
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Robots-Tag: none
X-Download-Options: noopen
X-Permitted-Cross-Domain-Policies: none
Set-Cookie: ockbfur1449k=6s9cvkj57mkbr1fja2464nbmhc; path=/owncloud; HttpOnly
Set-Cookie: cookie_test=test; expires=Tue, 14-Nov-2017 11:53:15 GMT; Max-Age=3600
Last-Modified: Tue, 14 Nov 2017 10:50:35 GMT
ETag: 09b1235dfb6b4953f62a3587a9adaf48
Content-Length: 2348
Content-Disposition: attachment; filename*=UTF-8''1510656635; filename="1510656635"
Content-Disposition: attachment; filename="bacon.txt"
OC-ETag: 09b1235dfb6b4953f62a3587a9adaf48
Content-Type: text/plain;charset=UTF-8
```

**Comment by DeepDiver1975** [comment]
> I'm now getting contradicting content-disposition headers:

adjusted now as well - happens if debug is false

**Comment by PVince81** [comment]
Most evil one fails for revert:

1. Login as "admin"
1. Enable encryption with master-key mode: `occ app:enable encryption && occ encryption:enable && occ encryption:select-encryption-type master-key
`
1. Create a user "user1".
1. Create a SFTP external storage, system-wide, mounted to "/sftp"
1. Set "Available for" of the storage to the "admin" group
1. Enable sharing in the cog icon for this storage
1. Create a folder "/sftp/enc"
1. Create a file "/sftp/enc/bacon.txt"
1. Edit the file several times to create versions
1. Share "/sftp/enc" with another user "user1". 
1. Login as "user1"
1. Browse to the versions of the file "bacon.txt"
1. Roll back to an older version.

Expected: rollback succeeds.
Actual: yellow notification about rollback failure

```
{"reqId":"ILwArlvGmwyk7TMd7dBR","level":4,"time":"2017-11-14T11:37:34+00:00","remoteAddr":"127.0.0.1","user":"user1","app":"webdav","method":"COPY","url":"\/owncloud\/remote.php\/dav\/meta\/42\/v\/1510659078","message":"Exception: {\"Message\":\"\\\/sftp\\\/enc\\\/bacon.txt not found while trying to get owner\",\"Exception\":\"OCP\\\\Files\\\\NotFoundException\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/Filesystem.php(960): OC\\\\Files\\\\View->getOwner('\\\/sftp\\\/enc\\\/bacon...')\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_versions\\\/lib\\\/Storage.php(94): OC\\\\Files\\\\Filesystem::getOwner('\\\/sftp\\\/enc\\\/bacon...')\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_versions\\\/lib\\\/Storage.php(323): OCA\\\\Files_Versions\\\\Storage::getUidAndFilename('\\\/sftp\\\/enc\\\/bacon...')\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/Storage\\\/Common.php(739): OCA\\\\Files_Versions\\\\Storage::rollback('\\\/sftp\\\/enc\\\/bacon...', '1510659078')\\n#4 [internal function]: OC\\\\Files\\\\Storage\\\\Common->restoreVersion('enc\\\/bacon.txt', '1510659078')\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/Storage\\\/Wrapper\\\/Wrapper.php(501): call_user_func_array(Array, Array)\\n#6 [internal function]: OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->__call('restoreVersion', Array)\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/Storage\\\/Wrapper\\\/Wrapper.php(501): call_user_func_array(Array, Array)\\n#8 [internal function]: OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->__call('restoreVersion', Array)\\n#9 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/Storage\\\/Wrapper\\\/Wrapper.php(501): call_user_func_array(Array, Array)\\n#10 [internal function]: OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->__call('restoreVersion', Array)\\n#11 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/Storage\\\/Wrapper\\\/Wrapper.php(501): call_user_func_array(Array, Array)\\n#12 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/Meta\\\/MetaFileVersionNode.php(105): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->__call('restoreVersion', Array)\\n#13 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Meta\\\/MetaFile.php(98): OC\\\\Files\\\\Meta\\\\MetaFileVersionNode->copy('\\\/user1\\\/files\\\/en...')\\n#14 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/DAV\\\/CopyPlugin.php(79): OCA\\\\DAV\\\\Meta\\\\MetaFile->copy('\\\/user1\\\/files\\\/en...')\\n#15 [internal function]: OCA\\\\DAV\\\\DAV\\\\CopyPlugin->httpCopy(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#16 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#17 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:COPY', Array)\\n#18 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#19 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Server.php(256): Sabre\\\\DAV\\\\Server->exec()\\n#20 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#21 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(175): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#22 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/Files\\\/View.php\",\"Line\":1653,\"User\":\"user1\"}"}
```

Maybe the rollback code needs similar path magic like we used in the other PR.

**Comment by DeepDiver1975** [comment]
💩

File: apps/files_versions/tests/VersioningTest.php [hunk 20]
```diff
@@ -118,57 +118,6 @@ protected function tearDown() {
 		parent::tearDown();
 	}
 
-
-	public function testMoveFileIntoSharedFolderAsRecipient() {
```
Comment (DeepDiver1975) [inline apps/files_versions/tests/VersioningTest.php]: removed - access to source is checked on the meta node level

File: apps/files_versions/tests/VersioningTest.php [hunk 21]
```diff
@@ -430,106 +379,6 @@ public function testMoveFolder() {
 	}
 
 
-	public function testMoveFolderIntoSharedFolderAsRecipient() {
```
Comment (DeepDiver1975) [inline apps/files_versions/tests/VersioningTest.php]: removed - access to source is checked on the meta node level

File: apps/files_versions/tests/VersioningTest.php [hunk 22]
```diff
@@ -430,106 +379,6 @@ public function testMoveFolder() {
...
-		$v1Renamed = $versionsFolder1 . '/folder1/folder2/test.txt.v' . $t1;
-		$v2Renamed = $versionsFolder1 . '/folder1/folder2/test.txt.v' . $t2;
-
-		$this->assertTrue($this->rootView->file_exists($v1Renamed));
-		$this->assertTrue($this->rootView->file_exists($v2Renamed));
-
-		\OC::$server->getShareManager()->deleteShare($share);
-	}
-
-	public function testRenameSharedFile() {
```
Comment (DeepDiver1975) [inline apps/files_versions/tests/VersioningTest.php]: removed - access to source is checked on the meta node level

File: lib/private/Files/Meta/MetaFileVersionNode.php [hunk 23]
```diff
@@ -102,6 +103,9 @@ public function getContent() {
 	public function copy($targetPath) {
 		$target = $this->root->get($targetPath);
 		if ($target instanceof File && $target->getId() === $this->parent->getId()) {
+			if (!$target->isUpdateable()) {
```
Comment (DeepDiver1975) [inline lib/private/Files/Meta/MetaFileVersionNode.php]: this is where the write access is verified

File: lib/private/Preview.php [hunk 24]
```diff
@@ -1334,8 +1334,9 @@ public static function post_delete_versions($args) {
 	 */
 	public static function post_delete($args, $prefix = '') {
 		$path = Files\Filesystem::normalizePath($args['path']);
+		$user = isset($args['user']) ? $args['user'] : \OC_User::getUser();
```
Comment (DeepDiver1975) [inline lib/private/Preview.php]: the user is now part of the hook arguments to make sure the correct user is used

File: apps/files_versions/lib/Storage.php [hunk 25]
```diff
@@ -309,68 +309,53 @@ public static function renameOrCopy($sourcePath, $targetPath, $operation) {
...
-			$users_view = new View('/'.$uid);
-			$files_view = new View('/'. User::getUser().'/files');
 
-			if (!$files_view->isUpdatable($filename)) {
-				return false;
-			}
-
-			$versionCreated = false;
+	public static function restoreVersion($uid, $filename, $fileToRestore, $revision) {
+		if(\OCP\Config::getSystemValue('files_versions', Storage::DEFAULTENABLED) !== true) {
```
Comment (PVince81) [inline apps/files_versions/lib/Storage.php]: old check was against string "true" and you replaced it with boolean `true`. I hope this won't break existing configs.

Of course, I'd expect the original code to use booleans...

Comment (DeepDiver1975) [reply apps/files_versions/lib/Storage.php]: Storage::DEFAULTENABLED is true and not 'true' and the old code used == and not ===/!==

File: lib/private/Preview.php (hunk 24 above)
Comment (PVince81) [reply lib/private/Preview.php]: Wow, our code can be quite obscure sometimes. I was confused when seeing that you extended the rollback event to include user, but this method is called `post_delete`.

Turns out that hook registration in base.php is wiring the "rollback" event to the "post_delete_versions" method here.

**Comment by PVince81** [comment]
- [ ] BUG: encryption + share from an ext storage: as recipient, revert file to older version: preview does not update on-disk even after page refresh.

revert-preview


It does work for local files.

**Comment by DeepDiver1975** [comment]
for previews there is a follow up pr - #29319

**Comment by PVince81** [comment]
@DeepDiver1975 does the follow up solve the problem ? I had the impression that it's a hook related bug when using ext storage + share, since regular scenarios work already.

If you intend to fix it in that PR, please add it as checklist item there.

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 does the follow up solve the problem ? I had the impression that it's a hook related bug when using ext storage + share, since regular scenarios work already.
> 
> If you intend to fix it in that PR, please add it as checklist item there.

The whole preview generation will be changed - fixing anything in here is useless work.
As example: I'm currently integrating objectstore versioning based on this branch - no previews are displayed at all.

**Comment by PVince81** [comment]
Thanks for clarification. I wasn't aware that the DAV previews would also touch the lower levels.

Sounds good then.

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by DeepDiver1975** [comment]
THX

**Comment by PVince81** [comment]
@DeepDiver1975 can you backport ICopySource to stable10 so we can fix https://github.com/owncloud/core/issues/29561 ?

**Comment by cortho** [comment]
@DeepDiver1975 regarding the change of this method:
https://github.com/owncloud/core/pull/29257/files#diff-efc44835a2c59b3707c965b80a3d669e
will files_versions then have a new version in info.xml (currently 1.3.0)?

**Comment by DeepDiver1975** [comment]
> will files_versions then have a new version in info.xml (currently 1.3.0)?

I see no reason to bump the version. Are you asking to find out about the API change?

**Comment by cortho** [comment]
ok.
> Are you asking to find out about the API change?
yes, this was the idea behind
